model/minsky.cc

```diff
@@ -713,6 +726,96 @@ namespace minsky
        });
     canvas.requestRedraw();
   }
+  
+  /*
+  For using GSL non-linear least square fitting routines
+  patterned on first Large Nonlinear Least Squares Example.
+  See online documentation at https://www.gnu.org/software/gsl/doc/html/nls.html
+  */
+      
+namespace
+{	
+
+   int residualFunc(const gsl_vector * x, void *params,
+          gsl_vector * f)
+   {
+	if (params==NULL) return GSL_EBADFUNC;
+	try { 	  
+		
+	      auto pars=static_cast<OptParams*>(params);  
+	      
+          int err = GSL_SUCCESS;
+          auto n = ValueVector::stockVars.size();
+          vector<double> xx(n);
+          
+          memcpy(xx.data(),reinterpret_cast<const vector<double>*>(x),n);
+          
+          err=gsl_odeiv2_driver_apply(pars->d->driver, pars->t,numeric_limits<double>::max(),&xx[0]);
+	      
+		  // define residual vector function to be optimized 
+          for (size_t i = 0; i < n; i++)
+              gsl_vector_set (f, i, xx[i] - pars->xi[i]);  
+     }
+     catch  (std::exception& e) {
+		((Minsky*)params)->threadErrMsg=e.what();
+        return GSL_EBADFUNC; 
+	 }   
+	 return  GSL_SUCCESS;      
+                        
+   }
+   
+   static void errHandlerOpt(const char* reason, const char* file, int line, int gsl_errno) {
+         throw error("gsl: %s:%d: %s",file,line,reason);
+   } 
```

Comment:
I'm a bit dubious about this approach. This callback will be called from within GSL, and we shouldn't throw exceptions within GSL, as it's not exception safe.

Reply:
Okay, I need to think about moving it, because I don't think it belongs in class Minsky either. As you have suggested before, the whole optimization data structure will have to go into a pimpl pattern much as RKdata, which will also take care of the public/private method question you pose below.